### DISCUSSION TITLE
Having trouble understanding the output of train_output.trained_net(), especially the distributional arguments for GPVAR

### DISCUSSION THREAD

**Discussion by yitao-yu** [awslabs/gluonts#1642] [Q&A] [answered]
I've been looking at[ issue #605 ](https://github.com/awslabs/gluon-ts/issues/605) for extracting covariance matrixes for each time step from a trained GPVAR model. 

Some of the codes needs to be changed to fit the version of gluonts that I'm working on (0.6.7). 

Here are my codes: 

```
from gluonts.dataset.loader import TrainDataLoader, InferenceDataLoader,ValidationDataLoader
import mxnet as mx
from functools import partial
from gluonts.mx.batchify import batchify

data_loader = TrainDataLoader(
            dataset=testing,
            transform=train_output.transformation,
            num_batches_per_epoch = estimator.trainer.num_batches_per_epoch,
            batch_size=1,
            stack_fn = partial(batchify, ctx=mx.cpu(), dtype=estimator.dtype) # 
)

from gluonts.support.util import get_hybrid_forward_input_names
from itertools import islice

data_entry = next(iter(data_loader))

input_names = get_hybrid_forward_input_names(
                train_output.trained_net
            )

loss, likelihoods, *distr_args = train_output.trained_net(
    *[data_entry[k] for k in input_names]
)

distr = distr_output.dist_cls(rank = rank, mu = distr_args[0], D = distr_args[1],W = distr_args[2],dim =  estimator.target_dim)

Sigma = distr.variance.asnumpy()
```
The distr_args I get is liike: 
`[<NDArray 1x2x17 @cpu(0)>,  <NDArray 1x2x17 @cpu(0)>,  <NDArray 1x2x17x5 @cpu(0)>]`

The rank of the model was 5, dimension was 17. So I was actually expecting a 17x5 matrix(W), and two 1x17 vector(Diagonal vector for D and mu). Instead, somehow for each of them, I got an additional dimension. So the Sigma calculated also contains two elements. I cannot understand that. 

I also cannot really identify D and mu. But for one of them I got all negative numbers and for another I got all positive. So I guess considering math, the all-positive one was D? I also cannot understand how I can get an all-negative vector, and it's true for multiple time steps. 

Here's how I train my model. You can see that the prediction_length was 1, so I was actually expecting one Sigma calculated. (I also cannot find any parameters with a default value of 2)

```
dim = len(monthly_df.columns)

distr_output = LowrankGPOutput(
        rank=rank, dim=dim, sigma_init=1.0
    )

trainer = Trainer(epochs=1000,batch_size=64,num_batches_per_epoch=10,learning_rate = 0.05,weight_decay = 1e-10)

gpvar = GPVAREstimator(freq = '20D', prediction_length = 1, target_dim = len(monthly_df.columns), 
                       trainer = trainer, dropout_rate = 0.2,
                       cell_type  = 'gru', num_cells = 32, num_layers = 4, 
                       conditioning_length = 3, use_marginal_transformation= True,
                       distr_output = distr_output
                       )

train_output = gpvar.train_model(training)
```

Thanks in advance for help! 
-----------------------------------Update----------------------------
I'm looking at the `./gpvar/_network.py ` file. Here's what it says in coment of the `hybrid_forward `method in the `GPVARTrainingNetwork`: 

```
        Returns
        -------
        distr
            Loss with shape (batch_size, 1)
        likelihoods
            Likelihoods for each time step
            (batch_size, context + prediction_length, 1)
        distr_args
            Distribution arguments (context + prediction_length,
            number_of_arguments)
```
That makes sense. The addition of context length and prediction length is 2(each is one). 

I previously thought that when context_length = 1, the model simply just make predictions as it unrolls. So I don't actually quite understand what's going on here. 

-----------------------------------Update2----------------------------
Use `InferenceDataSet `and `predictor.prediction_net` for inference. That will not yield the extra output, but it will output a num_samples * dim ndarray that I cannot extract distr_args from. 

So the forward propagation of the training network of GPVAR actually utilizes the `train_hybrid_forward `method of Deepvar. 

By default, the context_length was equal to 1(the prediction length) for GPVAR. That explains the extra output. The actual prediction was the second one. 

Still making sense of this unrolling. I think it is related to the data augmentation strategy the paper talked about at the end of section 2. So the context length is actually T? 

```
        # unroll the decoder in "training mode", i.e. by providing future data
        # as well
        rnn_outputs, _, scale, lags_scaled, inputs = self.unroll_encoder(
            F=F,
            past_time_feat=past_time_feat,
            past_target_cdf=past_target_cdf,
            past_observed_values=past_observed_values,
            past_is_pad=past_is_pad,
            future_time_feat=future_time_feat,
            future_target_cdf=future_target_cdf,
            target_dimension_indicator=target_dimension_indicator,
        )
```

**Comment by mbohlkeschneider** [comment, marked as answer]
Sorry, I'm a bit lost here. Can you restate your question?

**Reply by yitao-yu** [reply]
Sure!

First of all thanks for being patient to read and reply. 

What I am confused about in this discussion is why when I'm training the model and loading the data into it, it will generate a context_length(default: equal to prediction length) + prediction length output. After reading the paper, I think it might be because the model unrolls on the time series for several steps before making the prediction. The paper referred to this as a data augmentation strategy. I don't understand why and how this works...

For the inference problem we discussed in another issue, I think I can provide some code in which I did inference and tried to extract covariance matrix from a Sample Output. 

```
    data_loader = InferenceDataLoader(
            dataset=testing,
            transform=train_output.transformation,
            batch_size=1,
            stack_fn = partial(batchify, ctx=mx.cpu(), dtype=estimator.dtype) # 
            )
    
    data_entry = next(iter(data_loader))
    
    input_names = get_hybrid_forward_input_names(
                predictor.prediction_net
            )
    
    output = predictor.prediction_net(*[data_entry[k] for k in input_names])
    
    samples = np.array([sample[0] for sample in output.asnumpy()[0]])
    
    Sigma = np.cov(samples.T)
    mu = np.array([np.mean(target) for target in samples.T])
```
